File: nodejs/Implementation/wrapper.js
'use strict';

var Driver = require('../src/driver.js');
var CardStructure = require('./cardStructure.js');

function Wrapper(config) {
  this._driver = new Driver(config);
  this._cardStructure = new CardStructure(this._driver);
}

Object.defineProperty(Wrapper.prototype, 'driver', {
  get: function() {
    return this._driver;
  }
});
/**
 * createOrModifyTarget()
 * Create a new Target or modify it if it already exists
 *
 * @param {Object} targetObject
 * @return {Promise} response
 */
Wrapper.prototype.createOrModifyTarget = function(targetObject) {
  var target;

  return this._normalizeTarget(targetObject)
  .then((targetNormalize) => {
    target = targetNormalize;
    return this._cardStructure.getUnicityId();
  })
  .then((unicityId) => {
    return this.driver.target.find(target[unicityId]);
  })
  .then((response) => {
    if (response.statusCode === 200) {
      return this.driver.target.update(response.body.id, target);
    }
  })
  .catch((error) => {
    if (error.statusCode === 404) {
      return this.driver.target.create(target);
    }
    throw error;
  });
};
/**
 * addTargetAndSendMessage()
 * Create a new Target or modify
 * it if it already exists and send a message to it
 *
 * @param {Object} targetObject
 * @param {String} actionId
 * @param {Object} messageContent
 * @return {Promise} response
 */
Wrapper.prototype.addTargetAndSendMessage = function(targetObject, actionId, messageContent) {
  return this.createOrModifyTarget(targetObject)
  .then((response) => {
    if (response.statusCode === 200) {
      var targetId = response.body.id;
      return this.driver.send.message(actionId, targetId, messageContent);
    }
    throw response;
  });
};
/**
 * createCampaign()
 * Create a new Campaign Action
 *
 * @param {Object} actionObject
 * @return {Promise} response
 */
Wrapper.prototype.createCampaign = function(actionObject) {
  return this.driver.action.create(actionObject);
};
/**
 * createSegment()
 * Create a new Segment
 *
 * @param {Object} segmentObject
 * @return {Promise} response
 */
Wrapper.prototype.createSegment = function(segmentObject) {
  return this.driver.segment.create(segmentObject);
};
/**
 * createTargetAndAddToSegments()
 * Create a new Target or modify
 * it if it already exists and add it to a segment
 *
 * @param {Object} targetObject
 * @param {String} segmentId
 * @return {Promise} response
 */
Wrapper.prototype.createTargetAndAddToSegment = function(targetObject, segmentId) {
  return this.createOrModifyTarget(targetObject)
  .then((response) => {
    if (response.statusCode === 200) {
      var targetId = response.body.id;
      return this.driver.segment.addToSegment(targetId, segmentId);
    }
    throw response;
  });
};
/**
 * validateCampaign()
 * Validate a Campaign Action
 *
 * @param {String} campaignId
 * @param {Object} validationObject
 * @return {Promise} response
 */
Wrapper.prototype.validateCampaign = function(campaignId, validationObject) {
  return this.driver.action.validate(campaignId, validationObject);
};
/**
 * testCampaign()
 * Test a Campaign Action
 *
 * @param {String} campaignId
 * @param {Object} testconfig
 * @return {Promise} response
 */
Wrapper.prototype.testCampaign = function(campaignId, testconfig) {
  return this.driver.action.test(campaignId, testconfig);
};
/**
 * createAndTestCampaign()
 * Create and Test a Campaign Action
 *
 * @param {Object} actionObject
 * @param {Object} testconfig
 * @return {Promise} response
 */
Wrapper.prototype.createAndTestCampaign = function(actionObject, testconfig) {
  return this.driver.action.create(actionObject)
  .then((response) => {
    if (response.statusCode === 200) {
      var targetId = response.body.id;
      return this.driver.action.test(targetId, testconfig);
    }
    throw response;
  });
};
/**
 * import()
 * Create and Execute an import
 *
 * @param {Object} importConfig
 * @param {String} filepath
 * @return {Promise} response
 */
Wrapper.prototype.import = function(importConfig, filepath) {
  return this.driver.import.create(importConfig)
  .then((response) => {
    if (response.statusCode === 200) {
      return this.driver.import.execute(response.body.id, filepath);
    }
    throw response;
  });
};

/**
* normalizeTarget() is a private method that convert field
* name to its ID value
*
* @param {Object} target
* @return {Promise} normalizetarget
*/
Wrapper.prototype._normalizeTarget = function(target) {
  return new Promise((resolve, reject) => {
    if (typeof target !== 'object') {
      reject(new TypeError('Parameter should be an object'));
    }
    var keys = Object.keys(target);
    var newTarget = {};

    this._cardStructure.getFieldsMap()
    .then(function(fields) {
      for (var i = 0, len = keys.length; i < len; ++i) {
        let key = keys[i];
        if (isNaN(key)) {
          newTarget[fields[key]] = target[key];
        } else {
          newTarget[key] = target[key];
        }
      }
      resolve(newTarget);
    });
  });
};

module.exports = Wrapper;

File: csharp/Driver/Implementation/Wrapper.cs
using RestSharp;
using System;
using System.Collections.Generic;
using NP6Api.Models;
using System.Net;
using Newtonsoft.Json.Linq;

namespace NP6Api.Implementation
{
    public class Wrapper
    {
        private Driver _driver;
        private string _unicity;
        private Dictionary<string, string> _fields;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="config">Configuration for the driver</param>
        public Wrapper(Dictionary<string, string> config)
        {
            _driver = new Driver(config);
            _unicity = null;
            _fields = new Dictionary<string, string>();
        }

        /// <summary>
        /// Create a segment
        /// </summary>
        /// <param name="segment">Segment you want to create</param>
        /// <returns>Return NResponse with StatusCode and Body</returns>
        public NResponse CreateSegment(SegmentModel segment)
        {
            return _driver.Segment.Create(segment);
        }

        /// <summary>
        /// Create a campaign
        /// </summary>
        /// <param name="campaign">Action you want to create</param>
        /// <returns>Return NResponse with StatusCode and Body</returns>
        public NResponse CreateCampaign(IActionModel campaign)
        {
            return _driver.Action.Create(campaign);
        }

        /// <summary>
        /// Create a campaign, then test it with giving test options
        /// </summary>
        /// <param name="campaign">Campaign you want to create</param>
        /// <param name="testOption">Tests parameters</param>
        /// <returns>Return NResponse with StatusCode and Body</returns>
        public NResponse CreateAndTestCampaign(IActionModel campaign, TestModel testOptions)
        {
            NResponse response = _driver.Action.Create(campaign);

            if (response.StatusCode != HttpStatusCode.OK)
            {
                return response;
            }
            JObject body = (JObject)response.Body;
            string campaignId = (string)body.GetValue("id");
            return _driver.Action.Test(campaignId, testOptions);
        }

        /// <summary>
        /// Create a target or update its value if it already exist
        /// </summary>
        /// <param name="target">Target you want to create (or update)</param>
        /// <returns>Return NResponse with StatusCode and Body</returns>
        public NResponse CreateOrModifyTarget(JsonObject target)
        {
            string unicityId = GetUnicityId();
            JsonObject normalizeTarget = NormalizeTarget(target);

            return _driver.Target.Create(normalizeTarget, (string)normalizeTarget[unicityId]);
        }

        /// <summary>
        /// Create a target (or update it) and send it a message
        /// </summary>
        /// <param name="target">Target you want to create (or update)</param>
        /// <param name="actionId">Message's id</param>
        /// <param name="message">Custom message content</param>
        /// <returns>Return NResponse with StatusCode and Body</returns>
        public NResponse AddTargetAndSendMessage(JsonObject target, string actionId, MessageModel message = null)
        {
            var response = CreateOrModifyTarget(target);

            if (response.StatusCode != HttpStatusCode.OK)
            {
                return response;
            }
            JObject body = (JObject)response.Body;
            string targetId = (string)body.GetValue("id");
            return _driver.Send.Message(actionId, targetId, message);
        }

        /// <summary>
        /// Create a target and add it to a segment
        /// </summary>
        /// <param name="target">Target you want to create</param>
        /// <param name="segmentId">Segment's id in which you want to add the target</param>
        /// <returns>Return NResponse with StatusCode and Body</returns>
        public NResponse CreateTargetAndAddToSegment(JsonObject target, string segmentId)
        {
            var response = CreateOrModifyTarget(target);
            if (response.StatusCode != HttpStatusCode.OK)
            {
                return response;
            }
            JObject body = (JObject)response.Body;
            string targetId = (string)body.GetValue("id");
            return _driver.Segment.AddToSegment(segmentId, targetId);
        }

        // Private methods

        /// <summary>
        /// Retrieve the unicity ID
        /// </summary>
        /// <returns>Unicity ID</returns>
        private string GetUnicityId()
        {
            if (_unicity == null)
            {
                var response = _driver.Field.List();

                if (response.StatusCode != HttpStatusCode.OK)
                {
                    throw new InvalidOperationException("Unable to retrieve all fields");
                }
                JArray fields = (JArray)response.Body;
                foreach (JObject field in fields.Children<JObject>())
                {
                    _fields[(string)field.GetValue("name")] = (string)field.GetValue("id");
                    if ((bool)field.GetValue("isUnicity"))
                    {
                        _unicity = (string)field.GetValue("id");
                    }
                }
                if (_unicity == null)
                {
                    throw new InvalidOperationException("You haven't set any unicity type");
                }
                return _unicity;
            }
            return _unicity;
        }

        /// <summary>
        /// Convert field name to its ID value
        /// </summary>
        /// <param name="target">Target to normalize</param>
        /// <returns>Normalized target</returns>
        private JsonObject NormalizeTarget(JsonObject target)
        {
            JsonObject newTarget = new JsonObject();
            int n;

            foreach (var field in target)
            {
                if (!int.TryParse(field.Key, out n))
                {
                    if (!_fields.ContainsKey(field.Key))
                    {
                        throw new ArgumentException("The field named " + field.Key + " doesn't exist");
                    }
                    newTarget.Add(_fields[field.Key], field.Value);
                }
                else
                {
                    newTarget.Add(field.Key, field.Value);
                }
            }
            return newTarget;
        }

        /// <summary>
        /// Validate a campaign
        /// </summary>
        /// <param name="id">Id of the campaign to validate</param>
        /// <param name="validate">Validation object</param>
        /// <returns>Return NResponse with StatusCode and Body</returns>
        public NResponse ValidateCampaign(string id, ValidationModel validate)
        {
            return _driver.Action.Validate(id, validate);
        }

        /// <summary>
        /// Create an Import and Execute it
        /// </summary>
        /// <param name="import">Import Object</param>
        /// <param name="filePath">File path</param>
        /// <returns>Return NResponse with StatusCode and Body</returns>
        public NResponse Import(ImportModel import, string filePath)
        {
            var response = _driver.Import.Create(import);
            if (response.StatusCode != HttpStatusCode.OK)
            {
                return response;
            }
            var body = (JObject)response.Body;
            var importId = (string)body.GetValue("id");
            return _driver.Import.Execute(importId, filePath);
        }
    }
}
